heat/core/tensor.py

```diff
@@ -1426,28 +1482,230 @@ def __repr__(self, *args):
+
+                        else:  # if the keys given ask for data across multiple nodes
+                            if len(list(self.__array[tuple(key)].shape)) > 1:  # handles the definition of the lshape
+                                lout = list(self.__array[tuple(key)].shape)
+
+                                lout[self.split if len(tuple(self.__array[tuple(key)].shape)) == len(self.gshape) else self.split - 1] = \
+                                    self.comm.allreduce(tuple(self.__array[tuple(key)].shape)[self.split if len(tuple(self.__array[tuple(key)].shape)
+                                                                                                                ) == len(self.gshape) else self.split - 1],
+                                                        MPI.SUM)
+
+                                return tensor(self.__array[tuple(key)], tuple(lout), self.dtype, self.split, self.device, self.comm)
+                            else:
+                                lout = self.comm.allreduce(tuple(self.__array[tuple(key)].shape)[0], MPI.SUM)
+                                return tensor(self.__array[tuple(key)], (lout, ), self.dtype, self.split, self.device, self.comm)
+
+                    # if there is no overlap then the output is all on one node:
+                    if len(list(self.__array[tuple(key)].shape)) > 1:
+                        lout = list(self.__array[tuple(key)].shape)
+
+                        lout[self.split if len(tuple(self.__array[tuple(key)].shape)) == len(self.gshape) else self.split - 1] = \
+                            self.comm.allreduce(tuple(self.__array[tuple(key)].shape)[self.split if len(tuple(self.__array[tuple(key)].shape)
+                                                                                                        ) == len(self.gshape) else self.split - 1], MPI.SUM)
+
+                        return tensor(self.__array[tuple(key)], tuple(lout), self.dtype, self.split, self.device, self.comm)
+                    else:
+                        lout = self.comm.allreduce(tuple(self.__array[tuple(key)].shape)[0], MPI.SUM)
+                        return tensor(self.__array[tuple(key)], (lout,), self.dtype, self.split, self.device, self.comm)
+
+                # if the given axes are not splits (must be ints for python)
+                elif key[self.split] in range(chunk_start, chunk_end):
+                    key = list(key)
+                    key[self.split] = key[self.split] - chunk_start
+                    try:
+                        return tensor(self.__array[tuple(key)], tuple(self.__array[tuple(key)].shape), self.dtype, self.split, self.device, self.comm)
+                    except ValueError:  # case of returning just one value
+                        return self.__array[tuple(key)]
+                else:
+
```

Comment:
Stepped slicing is currently bugged. Try this example:

```
import heat as ht

a = ht.zeros((15,3,), split=0)
b = a[::2]

print(b.lshape)
```

will result in (mpirun -np 3):

```
[1,2]<stderr>:Traceback (most recent call last):
[1,2]<stderr>:  File "foo.py", line 4, in <module>
[1,2]<stderr>:    b = a[::2]
[1,2]<stderr>:  File "/home/mgoetz/Documents/heat/heat/core/tensor.py", line 1590, in __getitem__
[1,2]<stderr>:    key_set = set(range(key.start, key.stop, key.step if key.step else 1))
[1,2]<stderr>:TypeError: 'NoneType' object cannot be interpreted as an integer
[1,1]<stderr>:Traceback (most recent call last):
[1,1]<stderr>:  File "foo.py", line 4, in <module>
[1,1]<stderr>:    b = a[::2]
[1,1]<stderr>:  File "/home/mgoetz/Documents/heat/heat/core/tensor.py", line 1590, in __getitem__
[1,1]<stderr>:    key_set = set(range(key.start, key.stop, key.step if key.step else 1))
[1,1]<stderr>:TypeError: 'NoneType' object cannot be interpreted as an integer
[1,0]<stderr>:Traceback (most recent call last):
[1,0]<stderr>:  File "foo.py", line 4, in <module>
[1,0]<stderr>:    b = a[::2]
[1,0]<stderr>:  File "/home/mgoetz/Documents/heat/heat/core/tensor.py", line 1590, in __getitem__
[1,0]<stderr>:    key_set = set(range(key.start, key.stop, key.step if key.step else 1))
[1,0]<stderr>:TypeError: 'NoneType' object cannot be interpreted as an integer
```

Reply:
fixed in latest commit. i thought i hit all the test cases but i guess i missed that one